### PR TITLE
Test maybeDeferred handling of a function that returns a Failure

### PR DESCRIPTION
**Pull request by wsanchez**
Test `maybeDeferred` handling of a function that returns a `Failure` (rather than raising an `Exception`).

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10094
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

File: src/twisted/test/test_defer.py [hunk 1]
```diff
@@ -680,6 +680,18 @@ def test_maybeDeferredSyncError(self):
         self.assertEqual(len(E), 1)
         self.assertEqual(str(E[0].value), expected)
 
+    def test_maybeDeferredSyncFailure(self):
+        """
+        L{defer.maybeDeferred} should catch exception raised by a synchronous
+        function and errback its resulting L{defer.Deferred} with it.
```
Comment (twm) [inline src/twisted/test/test_defer.py]: It looks like this docstring is the same as the one for `test_maybeDeferredSyncException` and doesn't match the body of the new test.

File: src/twisted/test/test_defer.py [hunk 2]
```diff
@@ -680,6 +680,18 @@ def test_maybeDeferredSyncError(self):
...
+        """
+        S, E = [], []
+        d = defer.maybeDeferred((lambda x: defer.fail(x + 5)), 10)
```
Comment (twm) [inline src/twisted/test/test_defer.py]: I don't think that this matches the PR description. The `lambda` returns a `Deferred`, not a `Failure`.

Additionally, on Python 3 exceptions must derive from `BaseException`. `Failure(15)` creates a broken `Failure` object:

```
>>> Failure(15).raiseException()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/twm/dev/twisted/src/twisted/python/failure.py", line 475, in raiseException
    raise self.value.with_traceback(self.tb)
AttributeError: 'int' object has no attribute 'with_traceback'
```

Perhaps this is a defect in `Failure`? It should probably coerce to `TypeError` internally, as you get if you try to raise a non-exception:

```
>>> raise 1
Traceback (most recent call last):
  File "<console>", line 1, in <module>
TypeError: exceptions must derive from BaseException
```

**Review by twm — CHANGES_REQUESTED** [review]
I think that this needs some work. It doesn't look like it covers the intended case. Thank you for taking this on, though! We should definitely have coverage of this case.

File: src/twisted/test/test_defer.py (hunk 1 above)
Comment (wsanchez) [reply src/twisted/test/test_defer.py]: whoops

File: src/twisted/test/test_defer.py (hunk 2 above)
Comment (wsanchez) [reply src/twisted/test/test_defer.py]: Huh, right, ok. Rewriting.

File: src/twisted/test/test_defer.py [hunk 3]
```diff
@@ -658,15 +658,27 @@ def test_maybeDeferredSync(self):
...
+        d = defer.maybeDeferred(lambda: result)
+        d.addCallbacks(S.append, E.append)
+        self.assertEqual(E, [])
+        self.assertEqual(len(S), 1)
+        self.assertIdentical(S[0], result)
+
+    def test_maybeDeferredSyncWithArgs(self):
+        """
+        L{defer.maybeDeferred} should pass arguments to the called function.
+        """
```
Comment (wsanchez) [inline src/twisted/test/test_defer.py:672]: I added a new test to make sure that the identical returned object is passed through, and kept the existing test for checking that arguments are passes to the called function.

File: src/twisted/test/test_defer.py (hunk 1 above)
Comment (wsanchez) [reply src/twisted/test/test_defer.py]: fixed.

File: src/twisted/test/test_defer.py (hunk 2 above)
Comment (wsanchez) [reply src/twisted/test/test_defer.py]: done

**Review by twm — APPROVED** [review]
Looks great. Thank you!

**Comment by wsanchez** [comment]
`Run tests / testing (3.7)` failed, but that looks completely unrelated:

```console
2021-03-12T18:07:45.4460536Z [FAIL]
2021-03-12T18:07:45.4460847Z Traceback (most recent call last):
2021-03-12T18:07:45.4462216Z   File "/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.7/site-packages/twisted/python/test/test_versions.py", line 129, in test_strWithPrerelease
2021-03-12T18:07:45.4463223Z     str(Version("dummy", 1, 0, 0, prerelease=1)), "[dummy, version 1.0.0rc1]"
2021-03-12T18:07:45.4464797Z   File "/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.7/site-packages/twisted/trial/_synctest.py", line 424, in assertEqual
2021-03-12T18:07:45.4465837Z     super().assertEqual(first, second, msg)
2021-03-12T18:07:45.4466510Z   File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/unittest/case.py", line 852, in assertEqual
2021-03-12T18:07:45.4467111Z     assertion_func(first, second, msg=msg)
2021-03-12T18:07:45.4467835Z   File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/unittest/case.py", line 1233, in assertMultiLineEqual
2021-03-12T18:07:45.4468715Z     self.fail(self._formatMessage(msg, standardMsg))
2021-03-12T18:07:45.4469704Z twisted.trial.unittest.FailTest: '[dummy, version 1.0.0.rc1]' != '[dummy, version 1.0.0rc1]'
2021-03-12T18:07:45.4470431Z - [dummy, version 1.0.0.rc1]
2021-03-12T18:07:45.4470818Z ?                      -
2021-03-12T18:07:45.4471118Z + [dummy, version 1.0.0rc1]
2021-03-12T18:07:45.4471323Z 
2021-03-12T18:07:45.4471485Z 
2021-03-12T18:07:45.4472124Z twisted.python.test.test_versions.VersionsTests.test_strWithPrerelease
```

@adiroiban this seems like something where you might know what's up?

**Comment by adiroiban** [comment]
@wsanchez ... this looks line an issue generated by an `incremental` update.

**Comment by adiroiban** [comment]
@wsanchez I have merged it with trunk and push the changes... hope you don't mind. This should fix the test.
Don't forget to pull the branch before you do more changes.

**Pull request merged**